File: 220.存在重复元素-iii.py
#
# @lc app=leetcode.cn id=220 lang=python3
#
# [220] 存在重复元素 III
#

# @lc code=start
class Solution:
    def containsNearbyAlmostDuplicate(self, nums: List[int], k: int, t: int) -> bool:
        # 第一反应 桶排序
        # Min = min(nums)
        # Max = max(nums)
        # if t==0:


        # 要放入的桶中已经有其他元素了，这时将nums[i]放进去满足差值<=t
        # 可能存在前面一个桶的元素并且与nums[i]的差值<=t 或者 存在后面一个桶的元素并且与nums[i]的差值<=t



        if t < 0 or k < 0:
            return False

        buckets_size = t + 1
        buckets = {}

        for i in range(len(nums)):
            buckets_num = nums[i] // buckets_size
            if buckets_num in buckets:
                return True
            
            buckets[buckets_num] = nums[i]

            # 检查前后
            if (buckets_num-1) in buckets and abs(buckets[buckets_num-1] - nums[i]) <= t:
                return True
            if (buckets_num+1) in buckets and abs(buckets[buckets_num+1] - nums[i]) <= t:
                return True

            if i >= k:
                buckets.pop(nums[i-k]//buckets_size)
        return False




# @lc code=end

File: 220.存在重复元素-iii.java
/*
 * @lc app=leetcode.cn id=220 lang=java
 *
 * [220] 存在重复元素 III
 */

// @lc code=start
class Solution {
    public boolean containsNearbyAlmostDuplicate(int[] nums, int k, int t) {
            // if(k<0 && t<0){
            //     return false;
            // }

            // Map<Long, Long> buckets = new HashMap<>();
            // long buckets_size = t + 1;
            // for(int i=0;i<nums.length;i++){
            //     long buckets_num = nums[i] / buckets_size;
            //     if(buckets.containsKey(buckets_num)) return true;
            //     if(buckets.containsKey(buckets_num+1)&&abs(nums[i]-buckets.get(buckets_num+1))<=t){
            //         return true;
            //     }
            //     if(buckets.containsKey(buckets_num-1)&&abs(nums[i]-buckets.get(buckets_num-1))<=t){
            //         return true;
            //     }
            //     buckets.put(buckets_num, (long)nums[i]);

            //     if(i>=k){
            //         long tmp = nums[i-k]/buckets_size;
            //         buckets.remove(tmp);
            //     }
            // }
            // return false;
            if (t < 0) return false;
        Map<Long, Long> d = new HashMap<>();
        long w = (long)t + 1;
        for (int i = 0; i < nums.length; ++i) {
            long m = getID(nums[i], w);
            // check if bucket m is empty, each bucket may contain at most one element
            if (d.containsKey(m))
                return true;
            // check the nei***or buckets for almost duplicate
            if (d.containsKey(m - 1) && Math.abs(nums[i] - d.get(m - 1)) < w)
                return true;
            if (d.containsKey(m + 1) && Math.abs(nums[i] - d.get(m + 1)) < w)
                return true;
            // now bucket m is empty and no almost duplicate in nei***or buckets
            d.put(m, (long)nums[i]);
            if (i >= k) d.remove(getID(nums[i - k], w));
        }
        return false;
    }


    private long getID(long x, long w) {
        return x < 0 ? (x + 1) / w - 1 : x / w;
    }





    public long abs(long a){
        if(a<0) return -a; return a;

    }
}

// @lc code=end

